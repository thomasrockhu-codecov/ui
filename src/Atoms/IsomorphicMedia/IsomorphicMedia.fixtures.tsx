import React from 'react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from '../../index';
import { IsomorphicMedia } from './IsomorphicMedia';

export const Element = () => (
  <ThemeProvider theme={createTheme()}>
    <>
      <IsomorphicMedia query={t => t.media.greaterThan(t.breakpoints.sm)}>gt sm</IsomorphicMedia>
      <IsomorphicMedia query={t => t.media.lessThan(t.breakpoints.sm)}>lt sm</IsomorphicMedia>
    </>
  </ThemeProvider>
);

export const HTML_FROM_SSR =
  '<div class="IsomorphicMedia__StyledDiv-qvvp3f-0 fjZVrq">gt sm</div><div class="IsomorphicMedia__StyledDiv-qvvp3f-0 difbEy">lt sm</div>';
