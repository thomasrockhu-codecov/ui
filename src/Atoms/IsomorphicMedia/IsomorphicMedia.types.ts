import { Theme } from '../../theme/theme.types';

export type Props = {
  as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
  query: string | ((theme: Theme) => string);
};
