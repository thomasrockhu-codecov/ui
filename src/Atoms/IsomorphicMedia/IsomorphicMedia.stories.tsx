import React from 'react';

import { IsomorphicMedia } from './IsomorphicMedia';

export default {
  title: 'Atoms / IsomorphicMedia',
  parameters: {
    component: IsomorphicMedia,
  },
};

export const basicUsage = () => {
  return (
    <>
      <IsomorphicMedia query={(t) => t.media.lessThan(t.breakpoints.md)}>
        I am only shown on screens smaller than medium size
      </IsomorphicMedia>
      <IsomorphicMedia query={(t) => t.media.between(t.breakpoints.md, t.breakpoints.lg)}>
        I am only shown between medium and large sizes
      </IsomorphicMedia>
      <IsomorphicMedia query={(t) => t.media.greaterThan(t.breakpoints.lg)}>
        I am only shown on screens bigger than large size
      </IsomorphicMedia>
    </>
  );
};

basicUsage.story = {
  name: 'Basic usage',
};

export const withSizeKey = () => {
  return (
    <>
      <IsomorphicMedia query={(t) => t.media.lessThan(t.breakpoints.md.size)}>
        I am only shown on screens smaller than medium size
      </IsomorphicMedia>
      <IsomorphicMedia query={(t) => t.media.between(t.breakpoints.md.size, t.breakpoints.lg.size)}>
        I am only shown between medium and large sizes
      </IsomorphicMedia>
      <IsomorphicMedia query={(t) => t.media.greaterThan(t.breakpoints.lg.size)}>
        I am only shown on screens bigger than large size
      </IsomorphicMedia>
    </>
  );
};

withSizeKey.story = {
  name: 'With size key',
};
