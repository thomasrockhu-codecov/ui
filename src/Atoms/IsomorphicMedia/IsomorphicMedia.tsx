import React from 'react';
import styled, { ThemeContext } from 'styled-components';
import R from 'ramda';
import useSSR from 'use-ssr';

import { Props } from './IsomorphicMedia.types';
import { Theme } from '../../theme/theme.types';

const negateMedia = R.pipe(R.split(' '), R.insert(1, 'not all and'), R.join(' '));

const StyledDiv = styled.div<{ query: Props['query'] }>`
  ${(p) => negateMedia(typeof p.query === 'string' ? p.query : p.query(p.theme))} {
    display: none;
  }
`;

const useIsomorphicMedia = (query: string | ((t: Theme) => string)) => {
  const theme = React.useContext(ThemeContext);
  const [matches, setMatches] = React.useState<boolean | null>(null);
  const mediaQuery = (typeof query === 'string' ? query : query(theme)).replace('@media ', '');

  const { isServer } = useSSR();
  const useIsomorphicLayoutEffect = !isServer ? React.useLayoutEffect : React.useEffect;

  // Effect won't run during SSR
  useIsomorphicLayoutEffect(() => {
    if (typeof window.matchMedia !== 'function') {
      return undefined;
    }

    const media = window.matchMedia(mediaQuery);

    // Set matches first time
    setMatches(media.matches);
    const handler = (e: MediaQueryListEvent) => setMatches(e.matches);

    // Then update matches
    // if screen hits new breakpoint
    media.addListener(handler);

    // And of course unsubscribe onunmount
    return () => media.removeListener(handler);
  }, [mediaQuery]);

  return matches;
};

const IsomorphicMedia: React.FunctionComponent<Props> = (props) => {
  const { isServer } = useSSR();
  const As = props.as || 'div';
  const matches = useIsomorphicMedia(props.query);
  const componentProps = R.omit(['query', 'as'], props);
  // matches === null implies SSR or first client render, also need isServer to verify SSR.
  // Show css fallback in SSR
  if (isServer && matches === null) return <StyledDiv {...props} />;
  return matches ? <As {...componentProps}>{props.children}</As> : null;
};

export { IsomorphicMedia, useIsomorphicMedia };
