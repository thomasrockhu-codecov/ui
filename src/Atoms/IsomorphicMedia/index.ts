import { IsomorphicMedia, useIsomorphicMedia } from './IsomorphicMedia';

export { useIsomorphicMedia };
export default IsomorphicMedia;
