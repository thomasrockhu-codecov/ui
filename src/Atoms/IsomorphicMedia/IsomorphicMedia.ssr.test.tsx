/**
 * @jest-environment node
 */

import React from 'react';
import { ThemeProvider } from 'styled-components';
import { renderToString } from 'react-dom/server';
import { createTheme } from '../../index';
import { Element, HTML_FROM_SSR } from './IsomorphicMedia.fixtures';
import { IsomorphicMedia, useIsomorphicMedia } from './index';

test('Server-side rendering', async () => {
  const html = renderToString(<Element />);

  expect(html).toBe(HTML_FROM_SSR);
});

test('useMedia hook: returns null if SSR', () => {
  const theme = createTheme();
  const ConsumerThatDoesntMatch = () => {
    const matches = useIsomorphicMedia(t => t.media.greaterThan(t.breakpoints.lg));
    return <div data-testid="consumer">{`${matches}`}</div>;
  };

  const html = renderToString(
    <ThemeProvider theme={theme}>
      <ConsumerThatDoesntMatch />
    </ThemeProvider>,
  );
  expect(html).toMatchSnapshot();
});

test('Server-side rendering works: rendering all medias with CSS workaround', () => {
  const html = renderToString(
    <ThemeProvider theme={createTheme()}>
      <>
        <IsomorphicMedia query={t => t.media.greaterThan(t.breakpoints.sm)}>
          This shows on and above sm
        </IsomorphicMedia>
        <IsomorphicMedia query={t => t.media.lessThan(t.breakpoints.sm)}>
          This shows below sm
        </IsomorphicMedia>
      </>
    </ThemeProvider>,
  );
  expect(html).toMatchSnapshot();
});
