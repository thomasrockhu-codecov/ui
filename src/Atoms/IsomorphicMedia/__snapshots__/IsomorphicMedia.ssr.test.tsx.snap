// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server-side rendering works: rendering all medias with CSS workaround 1`] = `"<div class=\\"IsomorphicMedia__StyledDiv-qvvp3f-0 fjZVrq\\">This shows on and above sm</div><div class=\\"IsomorphicMedia__StyledDiv-qvvp3f-0 difbEy\\">This shows below sm</div>"`;

exports[`useMedia hook: returns null if SSR 1`] = `"<div data-testid=\\"consumer\\">null</div>"`;
